columnflow/config_util.py

```diff
@@ -39,6 +50,130 @@ def get_root_processes_from_campaign(campaign: od.Campaign) -> od.UniqueObjectIn
     return index
 
 
+def get_datasets_from_process(
+    config: od.Config,
+    process: str | od.Process,
+    strategy: str = "inclusive",
+    only_first: bool = True,
+    check_deep: bool = False,
+) -> list[od.Dataset]:
+    r"""
+    Given a *process* and the *config* it belongs to, returns a list of order dataset objects that
+    containg matching processes. This is done by walking through *process* and its child processes
+    and checking whether their are contained in known datasets. *strategy* controls how possible
+    amgiguities are resolved:
+
+        - ``"all"``: The full process tree is traversed and all matching datasets are considered.
+            Note that this might lead to a potential overrepresentation of the phase space.
+        - ``"inclusive"``: If a dataset is found to match a process, its child processes are not
+            checked further.
+        - ``"exclusive"``: If **any** (deep) subprocess of *process* is found to be contained in a
+            dataset, return datasets of subprocesses but not that of *process* itself (if any).
+        - ``"exclusive_strict"``: If **all** (deep) subprocesses of *process* are found to be
+            contained in a dataset, return these datasets but not that of *process* itself (if any).
+
+    As an example, consider the process tree
+
+    .. code-block:: none
+
+               --- single_top ---
+              /        |         \
+             /         |          \
+        s_channel  t_channel  tw_channel
+           / \        / \
+          /   \      /   \
+         t   tbar   t   tbar
+
+    and datasets existing for
+
+    .. code-block:: none
+
+        1. single_top__s_channel_t
+        2. single_top__s_channel_tbar
+        3. single_top__t_channel
+        4. single_top__t_channel_t
+        5. single_top__tw_channel
+
+    in the *config*. Depending on *strategy*, the returned datasets for process ``single_top``are:
+
+        - ``"all"``: ``[1, 2, 3, 4, 5]``. Simply all datasets matching any subprocess.
+        - ``"inclusive"``: ``[1, 2, 3, 5]``. Skipping ``single_top__t_channel_t`` since a more
+            inclusive dataset (``single_top__t_channel``) exists.
+        - ``"exclusive"``: ``[1, 2, 4, 5]``. Skipping ``single_top__t_channel`` since a more
+            exclusive dataset (``single_top__t_channel_t``) exists.
+        - ``"exclusive_strict"``: ``[1, 2, 3, 5]``. Not skipping ``single_top__t_channel`` since not
+            all subprocesses of ``t_channel`` match a dataset (there is no
+            ``single_top__t_channel_tbar`` dataset).
+
+    In addition, two arguments configure how the check is performed whether a process is contained
+    in a dataset. If *only_first* is *True*, only the first matching dataset is considered.
+    Otherwise, all datasets matching a specific process are returned. Forthe check itself,
+    *check_deep* is forwarded to :py:meth:`order.Datasaet.has_process`.
```

Comment:
> Also, just checking that I understand:check_deep=True would thread down on the Dataset tree rather than Process, right? So a process P would be matched to a dataset D if a child dataset D' is marked as containing the process P, even if D itself is not marked as containing P?

The datasets are actually not nested, but the `check_deep` is really just forwarded to the `has_process()` call of the dataset itself.

Reply:
Yes, but what does `Dataset.has_process(..., deep=True)` actually do compared to `Dataset.has_process(..., deep=False)`? From the implementation I am guessing it behaves as I explained above, i.e. threading potential children of the `self` `Dataset` object, or did I get that wrong?

So if we actually don't make use of any parent-child relationships between `Dataset` objects, do we really need `check_deep` at all here?

Reply:
Ah, I see! Seems I indeed got it wrong the first time, but looking again at the implementation in `order` I see that the threading is actually over child `Process` objects, in which case it makes complete sense to keep the `check_deep` flag as-is. Sorry for the noise :)